risc0/cargo-risczero/src/commands/deploy.rs

```diff
@@ -0,0 +1,115 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+use bonsai_sdk::{alpha::Client, API_KEY_ENVVAR, API_URL_ENVVAR};
+use clap::Parser;
+use risc0_build::BuildStatus;
+
+use crate::commands::build_guest;
+
+/// Reports the current version of this crate.
+pub const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// `cargo risczero deploy`
+///
+/// NOTE: Requires Docker to be installed and running.
+#[derive(Parser)]
+pub struct DeployCommand {
+    /// Location of the Cargo.toml for the guest code.
+    ///
+    /// This path is relative to the current directory.
+    #[arg(long)]
+    pub manifest_path: PathBuf,
+
+    /// Feature flags passed to cargo.
+    #[arg(long, value_delimiter = ',')]
+    pub features: Vec<String>,
+
+    /// API URL for Bonsai.
+    #[arg(long)]
+    pub api_url: Option<String>,
+
+    /// API key for Bonsai.
+    #[arg(long)]
+    pub api_key: Option<String>,
+}
+
+impl DeployCommand {
+    pub fn run(&self) -> Result<()> {
+        // Instantiate client first to check for errors
+        let client = self.get_client()?;
+        self.deploy(client)?;
+
+        Ok(())
+    }
+
+    fn deploy(&self, client: Client) -> Result<()> {
+        // Ensure we have an up to date artifact before deploying
+        if let BuildStatus::Skipped = build_guest::build(&self.manifest_path, &self.features)? {
+            eprintln!("Build skipped, nothing to deploy.");
+            return Ok(());
+        }
+
+        let src_dir = std::env::current_dir().unwrap();
+        let root_pkg = risc0_build::get_root_pkg(&self.manifest_path, &src_dir)?;
+        for target in risc0_build::get_targets(&root_pkg) {
+            if target.is_bin() {
+                let elf_path = risc0_build::get_elf_path(&src_dir, &root_pkg.name, &target.name);
+                let elf = std::fs::read(&elf_path).with_context(|| {
+                    format!("Failed to read ELF file at path: {}", elf_path.display())
+                })?;
+                let image_id = risc0_binfmt::compute_image_id(&elf)?;
+                let image_id_hex = hex::encode(image_id);
+                client.upload_img(&image_id_hex, elf)?;
+                eprintln!(
+                    "Uploaded ELF `{}` with image ID `{}`.",
+                    target.name, image_id_hex
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_client(&self) -> Result<Client> {
```

Comment:
i can foresee this being moved somewhere more general if we create more commands that use bonsai API. Maybe we can move this to this file https://github.com/risc0/risc0/blob/40e38a7ddbdc6d3a479a5efde999c4c2bc3c1e6a/risc0/cargo-risczero/src/utils.rs?

Reply:
Thanks for the feedback!

I moved the `get_client` and `get_client_envs` functions to `utils.rs` - Also learned that you can `flatten` clap arguments so we're able to now leverage `ClientEnvs` in other places if we need `get_client` and it will be consistent. :)